File: src/main/java/org/pl/interpreter/symbol/SymbolTable.java
package org.pl.interpreter.symbol;

import java.util.HashMap;
import java.util.Map;

public class SymbolTable {

    private SymbolTable parent;
    private Map<String, Object> symbols = new HashMap<>();

    public SymbolTable(SymbolTable parent) {
        this.parent = parent;
    }

    public Object get(String identifier) {
        var result = symbols.get(identifier);
        if (parent != null && result == null) {
            return parent.get(identifier);
        }
        return result;
    }

    public void set(String identifier, Object value) {
        symbols.put(identifier, value);
    }

    public void remove(String identifier) {
        symbols.remove(identifier);
    }

    @Override
    public String toString() {
        return "SymbolTable{" +
                "parent=" + parent +
                ", symbols=" + symbols +
                '}';
    }
}

File: src/main/java/org/pl/symbol/SymbolTable.kt
package org.pl.symbol

/**
 * Implementation of the symbol table specification. The [parent] symbol table
 * is used to allow scoping with the symbol table.
 */
class SymbolTable(private val parent: ISymbolTable?) : ISymbolTable {

    private val symbols = hashMapOf<String, Symbol>()

    override fun set(identifier: String, symbol: Symbol) {
        symbols[identifier] = symbol
    }

    override fun get(identifier: String): Symbol? {
        val symbol = symbols[identifier]
        if (parent != null && symbol == null) {
            return parent.get(identifier)
        }
        return symbol
    }

    override fun remove(identifier: String) = symbols.remove(identifier)
    override fun toString() = "SymbolTable(parent=$parent, symbols=$symbols)"
}